File: python/src/discovery/validation.py
"""POST-HOC validation: compare BLIND discoveries against the paper (held-out test set).

This is the ONLY module permitted to read the paper catalog / reported [[n,k,d]]
(blind_discovery_policy). It runs strictly AFTER a blind discovery campaign and never
feeds anything back into the search. It parses the CSS catalog tables and a small set of
paper-referenced landmark codes, then classifies each discovery as a parameter MATCH,
UPPER-BOUND-CONSISTENT, or NOVEL-at-n. Paper anchor: arXiv:2606.02418 catalogs + Sec V.B. R2.
"""
from __future__ import annotations

import re
from pathlib import Path

from ..algebra.polynomials import parse_poly, normalize_terms


def _num(tok: str):
    """Extract (value:int|None, is_upper:bool) from a catalog k/d cell (handles \\leq, \\textbf)."""
    is_upper = ("leq" in tok) or ("<=" in tok) or ("\\leq" in tok)
    m = re.search(r"\d+", tok.replace(",", ""))
    return (int(m.group()) if m else None), is_upper


def parse_css_catalog(tex_path: str | Path) -> list[dict]:
    """Parse css_catalog_tables.tex rows into reported CSS codes (held-out reference)."""
    text = Path(tex_path).read_text()
    codes: list[dict] = []
    for raw in text.splitlines():
        line = raw.strip()
        if " & " not in line or not line.endswith("\\\\"):
            continue
        cols = [c.strip() for c in line[:-2].split("&")]
        latt = re.search(r"\(\s*(\d+)\s*,\s*(\d+)\s*\)", cols[1]) if len(cols) > 1 else None
        if not latt or len(cols) < 7:
            continue
        l, m = int(latt.group(1)), int(latt.group(2))
        k, _ = _num(cols[4])
        d, d_up = _num(cols[5])
        if k is None or d is None:
            continue
        codes.append({
            "l": l, "m": m, "n": 2 * l * m, "k": k, "d": d, "d_is_upper": d_up,
            "A": cols[2], "B": cols[3], "pattern": cols[10] if len(cols) > 10 else "",
            "source": "css_catalog",
        })
    return codes


def parse_pbb_catalog(tex_path: str | Path) -> list[dict]:
    """Parse pbb_catalog_tables.tex rows into reported non-CSS PBB codes (held-out reference).

    Columns: Cl & (l,m) & A & B & C & D & k & d & FOM.
    """
    text = Path(tex_path).read_text()
    codes: list[dict] = []
    for raw in text.splitlines():
        line = raw.strip()
        if " & " not in line or not line.endswith("\\\\"):
            continue
        cols = [c.strip() for c in line[:-2].split("&")]
        latt = re.search(r"\(\s*(\d+)\s*,\s*(\d+)\s*\)", cols[1]) if len(cols) > 1 else None
        if not latt or len(cols) < 9:
            continue
        l, m = int(latt.group(1)), int(latt.group(2))
        k, _ = _num(cols[6])
        d, d_up = _num(cols[7])
        if k is None or d is None:
            continue
        codes.append({"l": l, "m": m, "n": 2 * l * m, "k": k, "d": d, "d_is_upper": d_up,
                      "A": cols[2], "B": cols[3], "pattern": "PBB", "source": "pbb_catalog"})
    return codes


def landmark_codes() -> list[dict]:
    """Paper-referenced CSS landmark/validation codes (Bravyi gross-code family, Sec V.B)."""
    return [
        {"l": 12, "m": 6, "n": 144, "k": 12, "d": 12, "d_is_upper": False,
         "A": "y+y^2+x^3", "B": "y^3+x+x^2", "pattern": "gross", "source": "landmark:bravyi2024high"},
        {"l": 6, "m": 6, "n": 72, "k": 12, "d": 6, "d_is_upper": False,
         "A": "y+y^2+x^3", "B": "y^3+x+x^2", "pattern": "gross-family", "source": "landmark:bravyi2024high(MILP-validation)"},
    ]


def _poly_set(expr, l, m):
    try:
        return frozenset(normalize_terms(parse_poly(expr, l, m), l, m))
    except Exception:
        return None


def validate(discoveries: list[dict], catalog_tex: str | Path | None, kind: str = "css") -> dict:
    """Classify each blind discovery against the paper reference. Returns a report dict.

    kind: 'css' (landmarks + CSS catalog, polynomial-set matching) or 'pbb' (PBB catalog,
    (n,k,d)-only matching since PBB equivalence is local-Clifford-level).
    classification per discovery:
      - 'MATCH'            : a reference code at the same n with equal k and equal d (both exact).
      - 'UB_CONSISTENT'    : same (n,k); our d is an upper bound consistent with reference d.
      - 'POLY_MATCH'       : same (l,m) and identical (A,B) polynomial sets as a reference code (css only).
      - 'NOVEL_AT_N'       : no reference code at this n (e.g. CSS catalog starts at n=144).
    """
    if kind == "pbb":
        reference = parse_pbb_catalog(catalog_tex) if (catalog_tex and Path(catalog_tex).exists()) else []
    else:
        reference = list(landmark_codes())
        if catalog_tex and Path(catalog_tex).exists():
            reference += parse_css_catalog(catalog_tex)
    ref_by_n: dict[int, list] = {}
    for r in reference:
        ref_by_n.setdefault(r["n"], []).append(r)

    results = []
    for disc in discoveries:
        n, k, d = disc["n"], disc["k"], disc.get("d")
        d_exact = disc.get("exact", False)
        refs = ref_by_n.get(n, [])
        verdict, matched = "NOVEL_AT_N", None
        if not refs:
            verdict = "NOVEL_AT_N"
        else:
            disc_A = _poly_set(disc["A"], disc["l"], disc["m"]) if isinstance(disc.get("A"), str) else \
                frozenset(map(tuple, disc.get("A", [])))
            disc_B = _poly_set(disc["B"], disc["l"], disc["m"]) if isinstance(disc.get("B"), str) else \
                frozenset(map(tuple, disc.get("B", [])))
            for r in refs:
                if r["k"] != k:
                    continue
                rA = _poly_set(r["A"], r["l"], r["m"])
                rB = _poly_set(r["B"], r["l"], r["m"])
                poly_hit = (disc_A is not None and rA is not None and
                            {disc_A, disc_B} == {rA, rB})
                # Identical polynomials => the SAME code. POLY_MATCH regardless of our d-estimate
                # (a BP-OSD upper bound may overestimate the reference's true d).
                if poly_hit:
                    verdict, matched = "POLY_MATCH", r
                    break
                if d is not None and d == r["d"] and d_exact and not r["d_is_upper"]:
                    verdict, matched = "MATCH", r
                    break
                if d is not None and d <= r["d"]:        # our (possibly upper-bound) d consistent
                    verdict, matched = "UB_CONSISTENT", r
            if matched is None and refs:
                verdict = "PARAMS_NOT_IN_REF_AT_N"
        results.append({"discovery": f"[[{n},{k},{d}]]", "fom": disc.get("fom"),
                        "exact": d_exact, "verdict": verdict,
                        "matched_ref": (f"[[{matched['n']},{matched['k']},{matched['d']}{'(<=)' if matched['d_is_upper'] else ''}]]"
                                        f" {matched['source']}") if matched else None})
    summary = {}
    for r in results:
        summary[r["verdict"]] = summary.get(r["verdict"], 0) + 1
    return {"n_reference_codes": len(reference), "results": results, "summary": summary}

File: julia/src/discovery/validation.jl
# POST-HOC validation: classify BLIND discoveries against known landmark codes.
# Pure-Julia port of qcode_discovery/discovery/validation.py (the landmark/validate path).
#
# This module is the Julia analogue of the ONLY Python module permitted to look at the
# held-out paper reference. Here we DO NOT bundle the paper catalog (.tex parsing) — per the
# task we use the built-in landmark codes ONLY (the Bravyi gross-code family, Sec V.B). It
# classifies each post-hoc discovery as POLY_MATCH / MATCH / UB_CONSISTENT / NOVEL_AT_N /
# PARAMS_NOT_IN_REF_AT_N exactly as the Python `validate(..., catalog_tex=None, kind="css")`.
#
# Paper anchor: arXiv:2606.02418 Sec V.B (post-hoc, held-out validation only).

"""
    landmark_codes() -> Vector{NamedTuple}

Paper-referenced CSS landmark/validation codes (Bravyi gross-code family, Sec V.B).
Each entry: (l, m, n, k, d, d_is_upper, A, B, pattern, source). Mirrors the Python
`landmark_codes()` list exactly (same two codes, same fields/values).
"""
function landmark_codes()
    return [
        (l = 12, m = 6, n = 144, k = 12, d = 12, d_is_upper = false,
         A = "y+y^2+x^3", B = "y^3+x+x^2", pattern = "gross",
         source = "landmark:bravyi2024high"),
        (l = 6, m = 6, n = 72, k = 12, d = 6, d_is_upper = false,
         A = "y+y^2+x^3", B = "y^3+x+x^2", pattern = "gross-family",
         source = "landmark:bravyi2024high(MILP-validation)"),
    ]
end

"""
    _normalize_terms(terms, l, m) -> Vector{Tuple{Int,Int}}

Reduce monomial exponents mod (l, m), cancel terms that appear an even number of times
(mod-2 ring arithmetic), then return the sorted, deduped survivors. Port of
`qcode_discovery.algebra.polynomials.normalize_terms`. NOTE: this is needed because the
existing Julia `parse_terms` reduces exponents mod (l,m) but does NOT cancel mod 2; the
polynomial-set match in Python compares the mod-2-reduced sets.
"""
function _normalize_terms(terms, l::Int, m::Int)
    counts = Dict{Tuple{Int,Int},Int}()
    for (a, b) in terms
        key = (mod(a, l), mod(b, m))
        counts[key] = get(counts, key, 0) + 1
    end
    survivors = [t for (t, c) in counts if isodd(c)]
    return sort(survivors)
end

"""
    _poly_set(expr, l, m) -> Union{Set{Tuple{Int,Int}}, Nothing}

The normalized monomial set of a polynomial, or `nothing` on parse failure. `expr` may be a
polynomial string (parsed with `parse_terms`) or an iterable of `(a,b)` tuples. Mirrors the
Python `_poly_set` (which returns a frozenset, or None on exception).
"""
function _poly_set(expr, l::Int, m::Int)
    try
        terms = expr isa AbstractString ? parse_terms(expr, l, m) :
                [(Int(t[1]), Int(t[2])) for t in expr]
        return Set(_normalize_terms(terms, l, m))
    catch
        return nothing
    end
end

"""
    validate(discoveries; kind=:css) -> NamedTuple

Classify each blind discovery against the built-in landmark codes (held-out reference).
`discoveries` is a vector of NamedTuples/Dicts with keys `n, k, l, m, A, B` and optionally
`d`, `exact`, `fom`. `kind` accepts `:css` only here (the `:pbb` path requires the paper
catalog, which is deliberately NOT bundled — see module docstring); any other kind errors.

Classification per discovery (identical precedence to Python `validate`):
  - `POLY_MATCH`             : same (l,m) and identical (A,B) polynomial sets as a reference.
  - `MATCH`                  : same (n,k), our exact d equals an exact reference d.
  - `UB_CONSISTENT`          : same (n,k), our (upper-bound) d <= a reference d.
  - `NOVEL_AT_N`             : no reference code at this n.
  - `PARAMS_NOT_IN_REF_AT_N` : reference codes exist at this n but none match (k/d).

Returns `(n_reference_codes, results, summary)` where `results` is a Vector{NamedTuple}
`(discovery, fom, exact, verdict, matched_ref)` and `summary` is a Dict{String,Int} of
verdict counts. Mirrors the Python report dict shape.
"""
function validate(discoveries; kind::Symbol = :css)
    if kind != :css
        throw(ArgumentError("kind=$kind unsupported in the Julia port: only :css " *
                            "(built-in landmarks) is available; the :pbb path needs the " *
                            "paper catalog, which is intentionally not bundled."))
    end
    reference = landmark_codes()
    ref_by_n = Dict{Int,Vector{eltype(reference)}}()
    for r in reference
        push!(get!(ref_by_n, r.n, eltype(reference)[]), r)
    end

    getf(disc, key, default = nothing) =
        disc isa NamedTuple ? get(disc, key, default) : get(disc, key, default)

    results = NamedTuple[]
    for disc in discoveries
        n = getf(disc, :n)
        k = getf(disc, :k)
        d = getf(disc, :d)
        d_exact = getf(disc, :exact, false) === true
        l = getf(disc, :l)
        mm = getf(disc, :m)
        refs = get(ref_by_n, n, eltype(reference)[])

        verdict = "NOVEL_AT_N"
        matched = nothing
        if isempty(refs)
            verdict = "NOVEL_AT_N"
        else
            disc_A = _poly_set(getf(disc, :A), l, mm)
            disc_B = _poly_set(getf(disc, :B), l, mm)
            for r in refs
                r.k != k && continue
                rA = _poly_set(r.A, r.l, r.m)
                rB = _poly_set(r.B, r.l, r.m)
                poly_hit = (disc_A !== nothing && rA !== nothing &&
                            Set([disc_A, disc_B]) == Set([rA, rB]))
                # Identical polynomials => the SAME code. POLY_MATCH regardless of our
                # d-estimate (an upper bound may overestimate the reference's true d).
                if poly_hit
                    verdict, matched = "POLY_MATCH", r
                    break
                end
                if d !== nothing && d == r.d && d_exact && !r.d_is_upper
                    verdict, matched = "MATCH", r
                    break
                end
                if d !== nothing && d <= r.d                 # our (possibly UB) d consistent
                    verdict, matched = "UB_CONSISTENT", r
                end
            end
            if matched === nothing && !isempty(refs)
                verdict = "PARAMS_NOT_IN_REF_AT_N"
            end
        end

        matched_ref = matched === nothing ? nothing :
            "[[$(matched.n),$(matched.k),$(matched.d)$(matched.d_is_upper ? "(<=)" : "")]] " *
            matched.source
        push!(results, (discovery = "[[$n,$k,$d]]", fom = getf(disc, :fom),
                        exact = d_exact, verdict = verdict, matched_ref = matched_ref))
    end

    summary = Dict{String,Int}()
    for r in results
        summary[r.verdict] = get(summary, r.verdict, 0) + 1
    end
    return (n_reference_codes = length(reference), results = results, summary = summary)
end
